File: BonusServer/Services/RuleTrigger/BonusRule_1.cs
using BonusServer.Models;
using FunLobbyUtils;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace BonusServer.Services.RuleTrigger
{
    public class BonusRule_1 : BonusRule
    {
        class TriggeringCondition_1 : BonusRule.TriggeringCondition
        {
            // 中獎率
            float mWinRatio;
            public float WinRatio
            {
                get {  return mWinRatio; }
                set
                {
                    if (value > 0.0f)
                    {
                        mWinRatio = value;
                        // cal WinDivided
                        int curValue = 10;
                        while (mWinRatio * curValue < 1.0f)
                        {
                            curValue *= 10;
                        }
                        this.RandomNum = (int)(mWinRatio * curValue);
                        this.RandomDivided = curValue;
                    }
                }
            }
            public int RandomNum { get; protected set; }
            public int RandomDivided { get; protected set; }
            // 啟動開獎條件2: 贏分次數
            public float MinTotalBet { get; set; }

            public TriggeringCondition_1() : base()
            {
                this.WinRatio = 0.001f;
                this.MinTotalBet = 0;
            }

            public override JObject Settings()
            {
                JObject obj = new JObject();
                obj["ScoreInterval"] = this.ScoreInterval;
                obj["WinRatio"] = this.WinRatio;
                obj["MinTotalBet"] = this.MinTotalBet;
                obj["MinPay"] = this.MinPay;
                return obj;
            }
        }

        public BonusRule_1() : base()
        {
            this.RuleId = RULEID.Rule_1;

            this.Condition_A = new TriggeringCondition_1();
            this.Condition_B = new TriggeringCondition_1();
            this.Condition_CR = new TriggeringCondition_1();
        }

        public override void ParseSettings(WIN_TYPE winType, string content)
        {
            try
            {
                if (content == null) throw new Exception("ParseSettings got null content");

                TriggeringCondition_1? src = JsonConvert.DeserializeObject<TriggeringCondition_1>(content);
                switch (winType)
                {
                    case WIN_TYPE.WinA:
                        {
                            TriggeringCondition_1? condition = this.Condition_A as TriggeringCondition_1;
                            if (condition == null) throw new Exception("ParseSettings got null condition for A");
                            condition.ScoreInterval = src.ScoreInterval;
                            condition.WinRatio = src.WinRatio;
                            condition.MinTotalBet = src.MinTotalBet;
                            condition.MinPay = src.MinPay;
                        }
                        break;
                    case WIN_TYPE.WinB:
                        {
                            TriggeringCondition_1? condition = this.Condition_B as TriggeringCondition_1;
                            if (condition == null) throw new Exception("ParseSettings got null condition for B");
                            condition.ScoreInterval = src.ScoreInterval;
                            condition.WinRatio = src.WinRatio;
                            condition.MinTotalBet = src.MinTotalBet;
                            condition.MinPay = src.MinPay;
                        }
                        break;
                    case WIN_TYPE.WinCR:
                        {
                            TriggeringCondition_1? condition = this.Condition_CR as TriggeringCondition_1;
                            if (condition == null) throw new Exception("ParseSettings got null condition for CR");
                            condition.ScoreInterval = src.ScoreInterval;
                            condition.WinRatio = src.WinRatio;
                            condition.MinTotalBet = src.MinTotalBet;
                            condition.MinPay = src.MinPay;
                        }
                        break;
                }
            }
            catch (Exception ex)
            {
                Log.StoreMsg(ex.Message);
            }
        }

        public override void Collect(WIN_TYPE winType, CollectData data)
        {
            switch (winType)
            {
                case WIN_TYPE.WinA:
                    {
                        WinCollection_A.TotalBet += data.TotalBet;
                        WinCollection_A.WinA += data.WinA;
                        TriggeringCondition_1? condition = this.Condition_A as TriggeringCondition_1;
                        // trigger win as long as WinBonusBeginDT is 0
                        if (condition != null &&
                            condition.WinBonusBeginUtcTime.Ticks == 0)
                        {
                            if (WinCollection_A.WinA > 0 &&
                                WinCollection_A.TotalBet >= condition.MinTotalBet)
                            {
                                condition.WinBonusBeginUtcTime = DateTime.UtcNow;
                            }
                            else
                            {
                                condition.WinBonusBeginUtcTime = Utils.zeroDateTime();
                            }
                        }
                    }
                    break;
                case WIN_TYPE.WinB:
                    {
                        WinCollection_B.TotalBet += data.TotalBet;
                        WinCollection_B.WinB += data.WinB;
                        TriggeringCondition_1? condition = this.Condition_B as TriggeringCondition_1;
                        // trigger win as long as WinBonusBeginDT is 0
                        if (condition != null &&
                            condition.WinBonusBeginUtcTime.Ticks == 0)
                        {
                            if (WinCollection_B.WinB > 0 &&
                                WinCollection_B.TotalBet >= condition.MinTotalBet)
                            {
                                condition.WinBonusBeginUtcTime = DateTime.UtcNow;
                            }
                            else
                            {
                                condition.WinBonusBeginUtcTime = Utils.zeroDateTime();
                            }
                        }
                    }
                    break;
                case WIN_TYPE.WinCR:
                    {
                        WinCollection_CR.TotalBet += data.TotalBet;
                        WinCollection_CR.WinB += data.WinB;
                        TriggeringCondition_1? condition = this.Condition_CR as TriggeringCondition_1;
                        // trigger win as long as WinBonusBeginDT is 0
                        if (condition != null &&
                            condition.WinBonusBeginUtcTime.Ticks == 0)
                        {
                            if (WinCollection_CR.WinB > 0 &&
                                WinCollection_CR.TotalBet >= condition.MinTotalBet)
                            {
                                condition.WinBonusBeginUtcTime = DateTime.UtcNow;
                            }
                            else
                            {
                                condition.WinBonusBeginUtcTime = Utils.zeroDateTime();
                            }
                        }
                    }
                    break;
            }
        }

        protected override bool RandomWin(WIN_TYPE winType)
        {
            TriggeringCondition_1? condition = null;
            switch (winType)
            {
                case WIN_TYPE.WinA:
                    condition = this.Condition_A as TriggeringCondition_1;
                    break;
                case WIN_TYPE.WinB:
                    condition = this.Condition_B as TriggeringCondition_1;
                    break;
                case WIN_TYPE.WinCR:
                    condition = this.Condition_CR as TriggeringCondition_1;
                    break;
            }

            if (condition != null)
            {
                // random value to get win
                Random rand = new Random();
                int num = rand.Next(0, condition.RandomDivided);
                if (num <= condition.RandomNum) return true;
            }
            return false;
        }
    }
}

File: BonusServer_Node/services/ruleTrigger/bonusRule_1.js

import Utils from '../../utils/utils.js';
import Log from '../../utils/log.js';
import {TriggeringCondition, BonusRule} from './bonusRule.js';

class TriggeringCondition_1 extends TriggeringCondition
{
    // 中獎率
    #mWinRatio;
    #mRandomNum;
    #mRandomDivided;

    // 啟動開獎條件2: 贏分次數
    MinTotalBet;

    WinRatio()
    {
        return this.#mWinRatio;
    }

    setWinRatio(value)
    {
        if (value > 0.0)
        {
            this.#mWinRatio = value;
            // cal WinDivided
            let curValue = 10;
            while (this.#mWinRatio * curValue < 1.0)
            {
                curValue *= 10;
            }
            this.#mRandomNum = Math.floor(this.#mWinRatio * curValue);
            this.#mRandomDivided = curValue;
        }
    }

    RandomNum()
    {
        return this.#mRandomNum;
    }
    
    RandomDivided()
    {
        return this.#mRandomDivided;
    }

    constructor()
    {
        super();        
        this.#mWinRatio = 0.001;
        this.MinTotalBet = 0;
    }

    Settings()
    {
        const obj = {
            ScoreInterval: this.ScoreInterval,
            WinRatio: this.WinRatio,
            MinTotalBet: this.MinTotalBet,
            MinPay: this.MinPay
        }
        return obj;
    }
}

export default class BonusRule_1 extends BonusRule
{
    constructor()
    {
        super(BonusRule.RULEID.Rule_1);

        this._Condition_A = new TriggeringCondition_1();
        this._Condition_B = new TriggeringCondition_1();
        this._Condition_CR = new TriggeringCondition_1();
    }

    ParseSettings(winType, content)
    {
        try
        {
            if (content == null) throw new Error("ParseSettings got null content");

            const src = JSON.parse(content);
            switch (winType)
            {
                case BonusRule.WIN_TYPE.WinA:
                    {
                        if (!this._Condition_A) throw new Error("ParseSettings got null condition for A");
                        this._Condition_A.ScoreInterval = src.ScoreInterval;
                        this._Condition_A.setWinRatio(src.WinRatio);
                        this._Condition_A.MinTotalBet = src.MinTotalBet;
                        this._Condition_A.MinPay = src.MinPay;
                    }
                    break;
                case BonusRule.WIN_TYPE.WinB:
                    {
                        if (!this._Condition_B) throw new Error("ParseSettings got null condition for B");
                        this._Condition_B.ScoreInterval = src.ScoreInterval;
                        this._Condition_B.setWinRatio(src.WinRatio);
                        this._Condition_B.MinTotalBet = src.MinTotalBet;
                        this._Condition_B.MinPay = src.MinPay;
                    }
                    break;
                case BonusRule.WIN_TYPE.WinCR:
                    {
                        if (!this._) throw new Error("ParseSettings got null condition for CR");
                        this._Condition_CR.ScoreInterval = src.ScoreInterval;
                        this._Condition_CR.setWinRatio(src.WinRatio);
                        this._Condition_CR.MinTotalBet = src.MinTotalBet;
                        this._Condition_CR.MinPay = src.MinPay;
                    }
                    break;
            }
        }
        catch (err)
        {
            Log.storeMsg(err.message ? err.message : err);
        }
    }

    Collect(winType, collectData)
    {
        switch (winType)
        {
            case BonusRule.WIN_TYPE.WinA:
                {
                    this._WinCollection_A.TotalBet += collectData.TotalBet;
                    this._WinCollection_A.WinA += collectData.WinA;
                    const condition = this._Condition_A;
                    // trigger win as long as WinBonusBeginDT is 0
                    if (condition && condition.WinBonusBeginUtcTime.getTime() == 0)
                    {
                        if (this._WinCollection_A.WinA > 0 &&
                            this._WinCollection_A.TotalBet >= condition.MinTotalBet)
                        {
                            condition.WinBonusBeginUtcTime = new Date();
                        }
                        else
                        {
                            condition.WinBonusBeginUtcTime = Utils.zeroDateTime();
                        }
                    }
                }
                break;
            case BonusRule.WIN_TYPE.WinB:
                {
                    this._WinCollection_B.TotalBet += collectData.TotalBet;
                    this._WinCollection_B.WinB += collectData.WinB;
                    const condition = this._Condition_B;
                    // trigger win as long as WinBonusBeginDT is 0
                    if (condition && condition.WinBonusBeginUtcTime.getTime() == 0)
                    {
                        if (this._WinCollection_B.WinB > 0 &&
                            this._WinCollection_B.TotalBet >= condition.MinTotalBet)
                        {
                            condition.WinBonusBeginUtcTime = new Date();
                        }
                        else
                        {
                            condition.WinBonusBeginUtcTime = Utils.zeroDateTime();
                        }
                    }
                }
                break;
            case BonusRule.WIN_TYPE.WinCR:
                {
                    this._WinCollection_CR.TotalBet += collectData.TotalBet;
                    this._WinCollection_CR.WinB += collectData.WinB;
                    const condition = this._Condition_CR;
                    // trigger win as long as WinBonusBeginDT is 0
                    if (condition && condition.WinBonusBeginUtcTime.getTime() == 0)
                    {
                        if (this._WinCollection_CR.WinB > 0 &&
                            this._WinCollection_CR.TotalBet >= condition.MinTotalBet)
                        {
                            condition.WinBonusBeginUtcTime = new Date();
                        }
                        else
                        {
                            condition.WinBonusBeginUtcTime = Utils.zeroDateTime();
                        }
                    }
                }
                break;
        }
    }

    RandomWin(winType)
    {
        let condition = null;
        switch (winType)
        {
            case BonusRule.WIN_TYPE.WinA:
                condition = this._Condition_A;
                break;
            case BonusRule.WIN_TYPE.WinB:
                condition = this._Condition_B;
                break;
            case BonusRule.WIN_TYPE.WinCR:
                condition = this._Condition_CR;
                break;
        }

        if (condition)
        {
            const randomDivided = condition.RandomDivided();
            const randomNum = condition.RandomNum();
            // random value to get win
            const num = Math.floor(Math.random() * randomDivided);
            if (num <= randomNum) return true;
        }
        return false;
    }
}